Add documentation about Hooks (#2877)
conan-io/docs: reference/extensions/hooks.rst

.. _reference_extensions_hooks:

Hooks
-----

> The Conan hooks is a feature intended to extend the Conan functionalities and let users customize the client behavior at determined
> points.

Reviewer: ```suggestion
The Conan hooks is a feature intended to extend the Conan functionalities to perform certain orthogonal operations, like some quality checks, in different stages of a package creation process, like pre-build and post-build.
```

Hook structure
--------------

A hook is a Python function that will be executed at certain points of Conan workflow
to customize the client behavior without modifying the client sources or the recipe ones.

> In the :ref:`hooks reference <hooks_reference>` you can find the full list of hook functions

Reviewer: This would be already the hooks reference, just the current section?

Author: Done, it has been incorporated to this same file.

                conanfile.output.error(f"[REQUIRED ATTRIBUTES] Conanfile doesn't have '{field}'.
                                          It is recommended to add it as attribute.")

    def pre_source(conanfile):
        conanfile_content = load(conanfile, os.path.join(conanfile.recipe_folder, "conanfile.py"))
        if "def source(self):" in conanfile_content:
            test = ""
            valid_content = [".zip", ".tar", ".tgz", ".tbz2", ".txz"]
            invalid_content = ["git checkout master", "git checkout devel", "git checkout develop"]
            if "git clone" in conanfile_content and "git checkout" in conanfile_content:
                fixed_sources = True
                for invalid in invalid_content:
                    if invalid in conanfile_content:
                        fixed_sources = False
            else:
                fixed_sources = False
                for valid in valid_content:
                    if valid in conanfile_content:
                        fixed_sources = True

            if not fixed_sources:
                output.error(f"[IMMUTABLE SOURCES] Source files does not come from and immutable place. Checkout to a "

>                             "commit/tag or download a compressed source file for {conanfile.ref}")

Reviewer: All of this sounds like an example, not reference material. I am fine with keeping a very simple, minimal example here im the reference, but the "immutable sources" seems too much.

Author: removed the second example.

relative module and warn, error or even raise to abort the Conan client execution.

Other useful task where a hook may come handy are the upload and download actions. There are **pre** and **post** functions for every
download/upload as a whole and for fine download tasks such as recipe and package downloads/uploads.

For example they can be used to sign the packages (including a file with the signature) when the package is created and check that
signature every time they are downloaded.

.. code-block:: python
   :caption: *hook_package_signing.py*

    import os
    from conan.tools.files import save, load

    SIGNATURE = os.getenv("ENTERPRISE_CONAN_PKG_SIGNATURE", "default-signature")

    def post_package(conanfile):
        sign_path = os.path.join(conanfile.package_folder, ".sign")
        save(conanfile, sign_path, SIGNATURE)
        conanfile.output.info("Package signed successfully")

    def post_package_info(conanfile):
        sign_path = os.path.join(conanfile.package_folder, ".sign")
        content = load(conanfile, sign_path)
        if content != SIGNATURE:

>             conanfile.output.error(f"[CHECK SIGNATURE] The signature file '{sign_path}' has a different content than the expected one.")

Reviewer: Same, this seems example, not reference.
Furthermore, the signing example is not good anymore that we have a dedicated plugin for signing packages.

Author: Removed the second example.

.. _reference_extensions_hooks:

Hooks
-----

The Conan hooks is a feature intended to extend the Conan functionalities to perform certain orthogonal operations,
like some quality checks, in different stages of a package creation process, like pre-build and post-build.

Hook structure
--------------

A hook is a Python function that will be executed at certain points of Conan workflow
to customize the client behavior without modifying the client sources or the recipe ones.

> 
> Hooks can implement any functionality: it could be Conan commands, recipe interactions
> such as exporting or packaging, or interactions with the remotes.

Reviewer: ```suggestion
```
It is not very clear what it means, and hooks cannot implement commands, or packaging or interaction with remotes. Better remove it

Here is an example of a simple hook:

.. code-block:: python
   :caption: *hook_example.py*

>     import os

Reviewer: ```suggestion
```

    from conan.tools.files import load

    def pre_export(conanfile):
        for field in ["url", "license", "description"]:
            field_value = getattr(conanfile, field, None)
            if not field_value:
                conanfile.output.error(f"[REQUIRED ATTRIBUTES] Conanfile doesn't have '{field}'.
                                          It is recommended to add it as attribute.")

> This hook checks the recipe content prior to it being exported and prior to downloading the sources. Basically the
> ``pre_export()`` function checks the attributes of the ``conanfile`` object to see if there is an URL, a license and a description and if missing,

Reviewer: ```suggestion
This hook checks the recipe content prior to it being exported. Basically the
``pre_export()`` function checks the attributes of the ``conanfile`` object to see if there is an URL, a license and a description and if missing,
```

warns the user with a message through the ``conanfile.output``. This is done **before** the recipe is exported to the local cache.

Any kind of Python script can be executed. You can create global functions and call them from different hook functions, import from a
relative module and warn, error or even raise to abort the Conan client execution.

> Other useful task where a hook may come handy are the upload and download actions. There are **pre** and **post** functions for every
> download/upload as a whole and for fine download tasks such as recipe and package downloads/uploads.

Reviewer: ```suggestion
```

Not anymore, the upload/download hooks have been removed, and left only those connected to recipe methods

Importing from a module
-----------------------

The hook interface should always be placed inside a Python file with the name of the hook starting by *hook_* and with the extension *.py*.

> It also should be stored in the *~/.conan2/extensions/hooks* folder. However, you can use functionalities from imported modules if you have

Reviewer: ```suggestion
It also should be stored in the *<conan_home>/extensions/hooks* folder. However, you can use functionalities from imported modules if you have
```

them installed in your system or if they are installed with Conan:

.. code-block:: python
   :caption: hook_example.py

    import requests
    from conan.tools.files import replace_in_file

    def post_export(conanfile):
        cmakelists_path = os.path.join(conanfile.export_source_folder, "CMakeLists.txt")

>         replace_in_file(conanfile, cmakelists_path, "PROJECT(MyProject)", "PROJECT(MyProject LANGUAGES CXX)")

Reviewer: Modification of files patch-like in a hook seems not a good practice, I wouldn't suggest this as an example. Better something that warns or raise an error if some file is not in the export_source_folder.

Author: Done, replaced by post_package example, downloading a license file.

        r = requests.get('https://api.github.com/events')

You can also import functionalities from a relative module:

.. code-block:: text

    hooks
    ├── custom_module
    │   ├── custom.py
    │   └── __init__.py
    └── hook_printer.py

Inside the *custom.py* from my *custom_module* there is:

.. code-block:: python
   :caption: custom.py

    def my_printer(conanfile):
        conanfile.output.info("my_printer(): CUSTOM MODULE")

And it can be used in the hook importing the module, just like regular Python:

.. code-block:: python
   :caption: hook_printer.py

    from custom_module.custom import my_printer

    def pre_export(conanfile):
        my_printer(conanfile)

Hook interface
--------------

> Here you can see a complete example of all the hook functions available and the different parameters for each of them depending on the
> context:

Reviewer: ```suggestion
Here you can see a complete example of all the hook functions available:
```

        conanfile.output.info("Running after of executing source() method.")

    def pre_generate(conanfile):
        conanfile.output.info("Running before to execute generate() method.")

    def post_generate(conanfile):
        conanfile.output.info("Running after of executing generate() method.")

    def pre_build(conanfile):
        conanfile.output.info("Running before to execute build() method.")

    def post_build(conanfile):
        conanfile.output.info("Running after of executing build() method.")

    def pre_package(conanfile):
        conanfile.output.info("Running before to execute package() method.")

    def post_package(conanfile):
        conanfile.output.info("Running after of executing package() method.")

    def pre_package_info(conanfile):
        conanfile.output.info("Running before to execute package_info() method.")

    def post_package_info(conanfile):
        conanfile.output.info("Running after of executing package_info() method.")

Functions of the hooks are intended to be self-descriptive regarding to the execution of them. For example, the ``pre_package()`` function
is called just before the ``package()`` method of the recipe is executed.

> Function parameters
> -------------------

Reviewer: ```suggestion
```

All hook methods are filled only with the same single object:

- **conanfile**: It is a regular ``ConanFile`` object loaded from the recipe that received the Conan command. It has its normal attributes
  and dynamic objects such as ``build_folder``, ``package_folder``, ``output``, ``dependencies``,  ``options`` ...

Storage, activation and sharing
-------------------------------

> Hooks are Python files stored under *~/.conan2/extensions/hooks* folder and **their file name should start with hook_ and end with the

Reviewer: ```suggestion
Hooks are Python files stored under *<conan_home>/extensions/hooks* folder and **their file name should start with hook_ and end with the
```

.py extension**.

The activation of the hooks is done automatically once the hook file is stored in the hook folder.
In case storing in subfolders, it works automatically too.

To deactivate a hook, its file should be removed from the hook folder. There is no configuration which can deactivate but keep the file stored in hooks folder.

Hooks are considered part of the Conan client configuration and can be shared as usual with the `reference_commands_config` command.
However, they can also be managed in isolated Git repositories cloned into the *~/.conan2/extensions/hooks* folder:

.. code-block:: bash

    $ cd ~/.conan2/extensions/hooks
    $ git clone https://github.com/myuser/my-conan-hooks.git my_hooks

> This way you can easily change from one version to another.

Reviewer: Are we sure about this, and like this is a good practice? I wouldn't say so, better to remove it until we discuss it.
Because it can open the door to having everything, profiles, settings.yml, conf, etc also as git repositories cloned directly into the Conan cache.

Author: Removed.
